src/lib/support/FluentTreeObject.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/support/Span.h>
+
+#include <optional>
+
+namespace chip {
+
+/// represents a wrapper around a type `T` that contains internal
+/// `Span<...>` values of other sub-types. It allows searching within the container sub-spans
+/// to create new containers.
+///
+/// Use case is that we very often search within a tree, like "find-endpoint" + "find-cluster" + "find-attribute"
+/// and we generally only care if "does the last element exist or not"
+///
+/// General usage is for fluent-searching for things like:
+///
+///    FluentTreeObject container(somePointer);
+///
+///    const AcceptedCommandData * value =
+///           container
+///              .Find<ByEndpoint>(path.mEndpointId, mEndpointIndexHint)
+///              .Find<ByServerCluster>(path.mClusterId, mServerClusterHint)
+///              .Find<ByAcceptedCommand>(path.mCommandId, mAcceptedCommandHint)
+///              .Value();
+///
+/// Where a `ByFoo` structure looks like:
+///
+///    struct ByFoo {
+///      using Key  = int;                  // the KEY inside a type
+///      using Type = SomeContainerStruct;  // the values for a sub-search
```

Comment:
So here Type would be the type of some object that the container `Find<ByFoo>` is being called on has a list of?  Calling it SomeContainerStruct is a bit weird, because Type may _not_ in fact be a container (e.g. in the code above it's EmberAfDeviceType).

Reply:
I was unsure what to call it. In practice it somehow `contains Span<something>`

Reply:
updated comments.